cvmfs/sqlitemem.cc

```diff
@@ -0,0 +1,489 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#define __STDC_FORMAT_MACROS
+
+#include "cvmfs_config.h"
+#include "sqlitemem.h"
+
+#include <cassert>
+#include <cstddef>
+#include <cstring>
+#include <new>
+
+#include "smalloc.h"
+#include "util_concurrency.h"
+
+using namespace std;  // NOLINT
+
+namespace sqlite {
+
+uint32_t MemoryManager::MallocArena::GetSize(void *ptr) const {
+  assert(Contains(ptr));
+
+  ReservedBlockCtl *block_ctl = reinterpret_cast<ReservedBlockCtl *>(
+    reinterpret_cast<char *>(ptr) - sizeof(ReservedBlockCtl));
+  int32_t size = -(block_ctl->size);
+  assert(size > 1);
+  return size - sizeof(ReservedBlockCtl) - 1;
+}
+
+/**
+ * Creates a free block at the place of the reserved block ptr points into.
+ * The free block might need to be merged with adjacent lower and/or upper
+ * blocks.  In these cases, the corresponding blocks are removed from the list
+ * of available blocks.  Every allocated block has a predecessor and a
+ * successor in the arena.  The newly created free block is added to the end of
+ * the list of available blocks.
+ */
+void MemoryManager::MallocArena::Free(void *ptr) {
+  assert(Contains(ptr));
+
+  no_reserved_--;
+
+  ReservedBlockCtl *block_ctl = reinterpret_cast<ReservedBlockCtl *>(
+    reinterpret_cast<char *>(ptr) - sizeof(ReservedBlockCtl));
+  char prior_tag = *(reinterpret_cast<char *>(block_ctl) - 1);
+  assert((prior_tag == kTagAvail) || (prior_tag == kTagReserved));
+  assert(block_ctl->size <= 0);
+
+  int32_t new_size = -(block_ctl->size);
+  assert(new_size > 0);
+  AvailBlockCtl *new_avail = reinterpret_cast<AvailBlockCtl *>(block_ctl);
+
+  if (prior_tag == kTagAvail) {
+    // Merge with block before and remove the block from the list
+    int32_t prior_size = reinterpret_cast<AvailBlockTag *>(
+      reinterpret_cast<char *>(block_ctl) - sizeof(AvailBlockTag))->size;
+    assert(prior_size > 0);
+    new_size += prior_size;
+    new_avail = reinterpret_cast<AvailBlockCtl *>(
+      reinterpret_cast<char *>(block_ctl) - prior_size);
+    AvailBlockCtl *next = new_avail->GetNextPtr(arena_);
+    AvailBlockCtl *prev = new_avail->GetPrevPtr(arena_);
+    prev->link_next = new_avail->link_next;
+    next->link_prev = new_avail->link_prev;
+    if (rover_ == new_avail)
+      rover_ = head_avail_;
+  }
+
+  int32_t succ_size = *reinterpret_cast<int32_t *>(
+    reinterpret_cast<char *>(new_avail) + new_size);
```

Comment:
This assumes that both `ReservedBlockCtl` and `AvailBlockCtl` have their size as the first member.